src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs

```diff
@@ -741,31 +743,33 @@ private ValueTask OnEndStreamReceived()
             }
         }
 
+        if (_context.WebTransportSession is not null)
```

Comment:
When the application code exists the WebTransportSession should also be gracefully closed.

Reply:
That's closing all sessions when the connection ends. You also need to close the specific session when it's original request ends. Somewhere around here:
https://github.com/dotnet/aspnetcore/blob/adefeb48ca0f11c365d44ed348a451dd4e2acc9c/src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs#L701

Reply:
Oh, I do that with this:
https://github.com/dotnet/aspnetcore/blob/14b026c792c6712a6c5eb23048dab59fe143c9ae/src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs#L52-L56

Reply:
However, I remember James mentioning that these callback registrations are expensive. Would your method be preferred over mine?

Reply:
That still seems to be tracking a closure of the Http3Connection, not the specific session. Look at where I linked to for Http3Stream cleanup logic.